github workflows: add basic CI workflow

weilianchu opened the pull request:
The first commit is a fix to a lint error caught by golintci.

This commit adds some basic automated workflows that trigger on every push and every PR against main branch.
It currently does three things:
 - runs all our go tests via call to `make test`. It does this test on ubuntu and macos.
 - runs `go mod tidy` and fails the job if there are any git diffs detected.
 - runs golangci-lint

Here's a run of this workflow on my own fork of the repo: https://github.com/weilianchu/leaderboard-backend/pull/1/checks

Closes https://github.com/speedrun-website/leaderboard-backend/issues/18

weilianchu committed:
github workflows: add basic CI workflow

This commit adds some basic automated workflows that trigger on every push and every PR against main branch.
It currently does two things:
 - runs all our go tests via call to make test
 - runs `go mod tidy` and `go fmt` and fails the job if there are any git diffs detected.

zdylag commented on .github/workflows/github-actions-demo.yml:
```diff
@@ -0,0 +1,26 @@
+name: leaderboard-backend CI
```
Discussion point for here and below: this name feels very redundant. Do other repos name these specific things?

Matt-Hurd replied:
Also file naming. This works and runs code, so it shouldn't be treated as a "demo".

FrozenSake replied:
Generally I believe actions get descriptive names such as "[build-test](https://github.com/actions/setup-go/blob/main/.github/workflows/workflow.yml)" or "[Licensed](https://github.com/actions/setup-go/blob/main/.github/workflows/licensed.yml)" or "[Validate 'setup-go'](https://github.com/actions/setup-go/blob/main/.github/workflows/versions.yml)" - all of which are examples from the github actions available actions.

weilianchu replied:
renamed to `backend-test` and `backend-lint` to be more useful

zdylag commented on .github/workflows/github-actions-demo.yml:
```diff
@@ -0,0 +1,26 @@
+
+jobs:
+
+  Golang-CI:
```
Same naming question.

weilianchu replied:
renamed to `backend-test` _some_ name needs to exist otherwise the actions just don't kick off

zdylag commented on .github/workflows/github-actions-demo.yml:
```diff
@@ -0,0 +1,26 @@
+jobs:
+
+  Golang-CI:
+    runs-on: ubuntu-latest
```
Ultra nit: I think this is the default it suggests, but I wonder if there is a lighter weight image that will save us on CI minutes.

weilianchu replied:
<img width="1083" alt="image" src="https://user-images.githubusercontent.com/9021608/125905530-7af0f88f-0ebe-4e80-92d2-d8e1fedddd11.png">
It seems just 10% of the runtime is spent on setting up the image, doesn't feel like it impacts our CI time much. Unless other images will perform better in other parts of the workflow as well?

FrozenSake replied:
Also github actions are [free for open source](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#about-billing-for-github-actions):

> GitHub Actions usage is free for both public repositories and self-hosted runners. For private repositories, each GitHub account receives a certain amount of free minutes and storage, depending on the product used with the account. Any usage beyond the included amounts is controlled by spending limits.

FrozenSake replied:
Of the actual [images](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources), there's nothing lighter weight without hosting our own. The list:
Windows Server 2019, Windows Server 2016
Ubuntu 20.04, Ubuntu 18.04
macOS Big Sur 11, macOS Catalina 10.15 

<!--EndFragment-->
</body>
</html>

Matt-Hurd replied:
I think the time right now is fine, as _most_ of it is bootstrapping time that shouldn't grow significantly with the codebase. 1 minute is fine, we can optimize once it starts growing and gets to be noticeable for our workflow.

zdylag reviewed (commented):
I left some very nit style questions. Mostly I don't feel comfortable approving since I don't know how to verify this does what we want.

@weilianchu would you mind adding a screenshot of the action from your repo (where it appears to have worked).

@Matt-Hurd did this not run since it is an unmerged action? If so do we expect it to work on all future PRs once merged?

weilianchu commented:
updated PR description with link to a test run of this workflow

Matt-Hurd commented:
> @Matt-Hurd did this not run since it is an unmerged action? If so do we expect it to work on all future PRs once merged?

From my experience, new actions aren't ran until merge.

Matt-Hurd commented on .github/workflows/github-actions-demo.yml:
```diff
@@ -0,0 +1,26 @@
+  Golang-CI:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checking out repository code
```
Nit: Verbiage. Rather than saying `Checking out repository code`, something like `Check out repository code`, or even shorten it to `Checkout repo`. No need to be so verbose.

weilianchu replied:
Sounds good! Changed to `Checkout repo`

FrozenSake commented on .github/workflows/backend-lint.yml:
```diff
@@ -0,0 +1,22 @@
+name: backend-lint
+
+on:
+  push:
+  pull_request:
+    branches: [ main ]
```
I think this is backwards. If we have
```suggestion
  push:
    branches: [ main ]
  pull_request:
```
then it only trigggers the push when the branch is main (force pushing to main), and otherwise the pull_request trigger handles synchronization events for in-PR pushes.

weilianchu replied:
ah yep confirmed this fixes my double action running!

FrozenSake commented on .github/workflows/backend-test.yml:
```diff
@@ -0,0 +1,21 @@
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  backend-test:
+    runs-on: ubuntu-latest
+    steps:
```
It would be nice to do a matrix here, since we're supporting local development on multiple OS, something like:

```suggestion
jobs:
  backend-test:
    strategy:
      matrix:
        go-version: [1.16.x]
        os: [macos-latest, ubuntu-latest, windows-latest]
    runs-on: ${{ matrix.os }}
    steps:
```
The go-version can be used below in set up golang to future proof supporting other version, just add them to the matrix.

FrozenSake commented on .github/workflows/backend-test.yml:
```diff
@@ -0,0 +1,21 @@
+name: backend-test
+
+on:
+  push:
+  pull_request:
+    branches: [ main ]
```
I think this is backwards. If we have
```suggestion
  push:
    branches: [ main ]
  pull_request:
```
then it only trigggers the push when the branch is main (force pushing to main), and otherwise the pull_request trigger handles synchronization events for in-PR pushes.

FrozenSake commented on .github/workflows/backend-test.yml:
```diff
@@ -0,0 +1,21 @@
+    - name: Set up Golang
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.16
```
This can use the matrix referenced above like so:
```suggestion
        go-version: ${{ matrix.go-version }}
```

FrozenSake reviewed (commented):
Overall, good solid actions. I think there's an issue with the push/pull_request limiting that may make it do the opposite of the intention, and I made a suggestion about using a matrix test for multi-OS compatibility.

For future linting I'm interested in testing out some stronger linters like [staticcheck ](https://staticcheck.io/)or [go fumpt](https://github.com/mvdan/gofumpt) but I think they are:
A) too heavyweight for an initial action
B) not tools I have enough experience with to pick between or recommend right now.

zdylag committed:
go middleware: check error for prometheus registrations in tests

There were a couple more registrations that went unchecked in our tests.

weilianchu committed:
go middleware: use ioutil.ReadAll instead of io.ReadAll

golangci-lint complained with: Error: ReadAll not declared by package io (typecheck)
Changing to ioutil did not break tests, and passed linter.

Matt-Hurd commented on .github/workflows/backend-lint.yml:
```diff
@@ -0,0 +1,45 @@
+
+    ## Run go mod tidy and then check if there was a diff. If so, fail the check.
+    - name: Check go mod tidy diff
+      run: if [ $(go mod tidy && git diff | wc -l) -gt 0 ]; then git diff && exit 1; fi
```
for those others curious about this line, there's no other good way of doing this: https://github.com/golang/go/issues/27005

Matt-Hurd reviewed (dismissed):
All comments have been addressed, I've resolved their conversations.

Edit: Nevermind, I missed something. The first commit currently in the PR is duplicated from #49. Can you remove it and ensure you're based off of main?

Edit 2: Also, sorry for the git noise. I'm still figuring this "admin" thing out 😅

weilianchu commented:
> Edit: Nevermind, I missed something. The first commit currently in the PR is duplicated from #49. Can you remove it and ensure you're based off of main?

Addressed! I cherrypicked that commit so zanzan could approve it and we could merge last night, but didn't get around to it haha.

zdylag reviewed (approved):
🎉 This looks great in your fork 🎉

FrozenSake reviewed (approved):
Agreed - this is a great set of actions to have! 🎉
